Challo/Challo/Modules/Chat/ChatPresenter.swift

```diff
@@ -0,0 +1,148 @@
+        guard let dialogId = currentOpenDialogId else {
+            fatalError("Attempting to send message without a current dialog open")
+        }
+        interactor.sendMessage(messageBody: messageText, dialogId: dialogId)
+        messageText = ""
+    }
+    
+    func onChatAppear() {
+        guard canDisplayChat else {
+            return
+        }
+        guard isChatAvailable else {
+            return
+        }
+        interactor.getDialogs()
+    }
+    
+    func formatDialogDatetime(_ date: Date) -> String {
+        if Calendar.current.isDateInToday(date) {
+            return CustomDateFormatter.displayDateAsTime(date)
+        } else {
+            let relativeDateTimeFormatter = RelativeDateTimeFormatter()
+            relativeDateTimeFormatter.dateTimeStyle = .numeric
+            return relativeDateTimeFormatter.localizedString(for: date, relativeTo: Date())
+        }
+    }
+}
+
+// MARK: UI Rendering Logic
+extension ChatPresenter {
+    private func makeMessageViewFromMessageIndex(_ index: Int, shouldDisplayAvatar: Bool = false) -> ChatMessageView {
+        let message = messages[index]
+        return ChatMessageView(message: message.message,
+                               isSentByCurrentUser: message.isSentByCurrentUser,
+                               isSuccessfullySent: message.isSuccessfullySent,
+                               shouldDisplayAvatar: shouldDisplayAvatar || index == messages.count - 1)
+    }
+    
+    private func transformPrevMessageViewToDisplayAvatar(messagesView: inout [AnyView], currentIndex: Int) {
+        messagesView.removeLast()
+        messagesView.append(AnyView(makeMessageViewFromMessageIndex(currentIndex - 1, shouldDisplayAvatar: true)))
+    }
+    
+    // Should append time when the current message is > 10 minutes apart from the prev message
+    private func shouldAppendDateTime(_ currentIndex: Int) -> Bool {
+        guard currentIndex != 0 else {
+            return true
+        }
+        guard let dateSent = messages[currentIndex].dateSent,
+              let prevDateSent = messages[currentIndex - 1].dateSent else {
+            return false
+        }
+        let diffInSeconds = dateSent.timeIntervalSince(prevDateSent)
+        return diffInSeconds > (60 * 10)
+    }
+    
+    private func appendDateTimeWhenNeeded(messagesView: inout [AnyView], currentIndex: Int,
+                                          hasTransformedPrevMessage: Bool) {
+        let message = messages[currentIndex]
+        guard let dateSent = message.dateSent, shouldAppendDateTime(currentIndex) else {
+            return
+        }
+        if !hasTransformedPrevMessage && currentIndex != 0 {
+            transformPrevMessageViewToDisplayAvatar(messagesView: &messagesView, currentIndex: currentIndex)
+        }
+        messagesView.append(AnyView(
+            Text(CustomDateFormatter.displayFriendlyDateTime(dateSent))
+        ))
+    }
+    
+    func makeMessagesView() -> some View {
```

Comment:
Should presenter be responsible for setting up views? Or should the main chat view be the one mapping presenter messages to messageviews?

Reply:
This rendering logic is more complex and needs to be done imperatively rather than declaratively, so it's left under the presenter